File: Packages/RallyCore/Sources/RallyCore/Models/CheckIn.swift
import Foundation

/// Represents a venue check-in with proof of attendance.
public struct CheckIn: Identifiable, Codable, Hashable, Sendable {
    public let id: String
    public let userID: String
    public let eventID: String
    public let venueID: String
    public let timestamp: Date
    public let proof: CheckInProof
    public let pointsEarned: Int
    public let status: CheckInStatus

    public init(
        id: String,
        userID: String,
        eventID: String,
        venueID: String,
        timestamp: Date = .now,
        proof: CheckInProof,
        pointsEarned: Int = 0,
        status: CheckInStatus = .pending
    ) {
        self.id = id
        self.userID = userID
        self.eventID = eventID
        self.venueID = venueID
        self.timestamp = timestamp
        self.proof = proof
        self.pointsEarned = pointsEarned
        self.status = status
    }
}

public struct CheckInProof: Codable, Hashable, Sendable {
    public let latitude: Double
    public let longitude: Double
    public let horizontalAccuracy: Double
    public let beaconUUID: String?
    public let beaconMajor: UInt16?
    public let beaconMinor: UInt16?
    public let beaconProximity: String?
    public let attestationToken: String?

    public init(
        latitude: Double,
        longitude: Double,
        horizontalAccuracy: Double,
        beaconUUID: String? = nil,
        beaconMajor: UInt16? = nil,
        beaconMinor: UInt16? = nil,
        beaconProximity: String? = nil,
        attestationToken: String? = nil
    ) {
        self.latitude = latitude
        self.longitude = longitude
        self.horizontalAccuracy = horizontalAccuracy
        self.beaconUUID = beaconUUID
        self.beaconMajor = beaconMajor
        self.beaconMinor = beaconMinor
        self.beaconProximity = beaconProximity
        self.attestationToken = attestationToken
    }
}

public enum CheckInStatus: String, Codable, Hashable, Sendable {
    case pending
    case verified
    case rejected
    case expired
}

File: Android/core/src/main/java/com/rally/app/core/model/CheckIn.kt
package com.rally.app.core.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Represents a venue check-in with proof of attendance.
 */
@Serializable
data class CheckIn(
    val id: String,
    val userID: String,
    val eventID: String,
    val venueID: String,
    val timestamp: Long = System.currentTimeMillis(),
    val proof: CheckInProof,
    val pointsEarned: Int = 0,
    val status: CheckInStatus = CheckInStatus.PENDING
)

/**
 * Location and beacon evidence used to verify a check-in.
 */
@Serializable
data class CheckInProof(
    val latitude: Double,
    val longitude: Double,
    val horizontalAccuracy: Double,
    val beaconUUID: String? = null,
    val beaconMajor: Int? = null,
    val beaconMinor: Int? = null,
    val beaconProximity: String? = null,
    val attestationToken: String? = null
)

/**
 * Verification status of a check-in.
 */
@Serializable
enum class CheckInStatus {
    @SerialName("pending") PENDING,
    @SerialName("verified") VERIFIED,
    @SerialName("rejected") REJECTED,
    @SerialName("expired") EXPIRED
}
